google/cloud/internal/oauth2_external_account_credentials.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/oauth2_external_account_credentials.h"
+#include "google/cloud/internal/external_account_parsing.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/internal/rest_client.h"
+#include <nlohmann/json.hpp>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+ExternalAccountCredentials::ExternalAccountCredentials(
+    ExternalAccountInfo info, HttpClientFactory client_factory, Options options)
+    : info_(std::move(info)),
+      client_factory_(std::move(client_factory)),
+      options_(std::move(options)) {}
+
+StatusOr<internal::AccessToken> ExternalAccountCredentials::GetToken(
+    std::chrono::system_clock::time_point tp) {
+  auto subject_token = (info_.token_source)(Options{});
+  if (!subject_token) return std::move(subject_token).status();
+
+  auto form_data = std::vector<std::pair<std::string, std::string>>{
+      {"grant_type", "urn:ietf:params:oauth:grant-type:token-exchange"},
+      {"requested_token_type", "urn:ietf:params:oauth:token-type:access_token"},
+      {"scope", "https://www.googleapis.com/auth/cloud-platform"},
+      {"audience", info_.audience},
+      {"subject_token_type", info_.subject_token_type},
+      {"subject_token", subject_token->token},
+  };
+  auto request = rest_internal::RestRequest(info_.token_url);
+
+  auto client = client_factory_(options_);
+  auto status = client->Post(request, form_data);
+  if (!status) return std::move(status).status();
+  auto response = *std::move(status);
+  if (MapHttpCodeToStatus(response->StatusCode()) != StatusCode::kOk) {
+    return AsStatus(std::move(*response));
+  }
+  auto payload = rest_internal::ReadAll(std::move(*response).ExtractPayload());
+  if (!payload) return std::move(payload.status());
```

Comment:
FYI: I am not the best reviewer of this chunk of code. Maybe @scotthart can give it the once-over?

Reply:
Ack.  There is a lot of similar code elsewhere